core/src/impl/Kokkos_Core.cpp

```diff
@@ -792,8 +792,14 @@ void post_initialize_internal(const Kokkos::InitializationSettings& settings) {
 }
 
 void initialize_internal(const Kokkos::InitializationSettings& settings) {
+  // Tools should only be used after the backends have been initialized (because
+  // they have not been initialized at that point) but if they have been set via
+  // set_callbacks they need to be present before calling post_initialize so
+  // that they get initialized as well.
```

Comment:
```suggestion
  // The tool initialization is only called in post_initialize_internal.
  // Pausing tools here, so that if someone has set callbacks programmatically
  // these callbacks are not called inside the backend initialization, before
  // the tool initialization happened.
```